### ISSUE TITLE
LLaMA-Factory Finetuning on RQ-VAE Codebook Prompts: Tokenization Issues and Missing Evaluation Pipeline

### ISSUE THREAD

**Issue by hupeiyu21** [wds1996/GNPR-SID#1]
Hi, thanks for the great work!

I'm currently trying to reproduce the codebook-based recommendation finetuning described in your paper/repo, using [LLaMA-Factory](https://github.com/hiyouga/LLaMA-Factory) as the backbone for instruction tuning.

### 🧪 Task Setup
I'm working with converted prompts where both the input and output are composed of codebook tokens generated by RQ-VAE, such as:

```

"prompt": "User\_123 visited: \<a\_112>\<b\_45>\<c\_22>, \<a\_76>\<b\_1>\<c\_94>, \<a\_176>\<b\_103>\<c\_179>",
"label": "\<a\_176>\<b\_103>\<c\_179>"

```

I’ve already ensured that the mapping between item_id and codebook is correct and that all special tokens (e.g., `<a_0>` to `<a_255>`) have been registered via `tokenizer.add_special_tokens()` and `model.resize_token_embeddings()`.

---

###  Problem 1: Incomplete Codebook Outputs

After finetuning with LLaMA-Factory, many generated sequences are **incomplete** or **truncated**, such as:

```

Input:  User\_123 visited: \<a\_20>\<b\_10>\<c\_60>, \<a\_31>\<b\_88>\<c\_45>
Output: \<a\_31>\<b\_88>

```

This breaks the codebook format and makes it impossible to map back to the original item.

---

### Question 1: How did you configure LLaMA-Factory for training?

Could you share your hyperparameters for LLaMA-Factory? For example:

- `model_name_or_path` = ?
- `cutoff_len`, `max_seq_len`, `train_on_inputs`, `add_special_tokens`, `eos_token_id`, etc.
- tokenizer config: Did you modify the tokenizer vocab files directly or register programmatically?

---

### Problem 2: Evaluation behavior (Hit@1 = 0.000)

I noticed that the evaluation script directly uses the predicted tokens to compute metrics. But in many cases, the generated token sequences **partially match** or include decoding noise.

I’d like to **save all predicted outputs first** and then run a post-processing script to evaluate matching (e.g., Hit@1) only if the entire predicted codebook exactly matches the label.

---

### Question 2: Could you share the evaluation strategy?

- How exactly did you evaluate the model’s predictions?
- Do you save intermediate generated results first?
- Did you do exact matching of codebook tokens (including `<a_...><b_...><c_...>`)?

---

### Summary

- I’d appreciate your training config and tokenizer setup for LLaMA-Factory
- I’d like to know your evaluation pipeline (prompt → prediction → metric)
- I suspect my low performance comes from partial decoding / tokenizer mismatch

Thanks a lot!
```

**Comment by wds1996** [comment]
**Q1**
You can find the hyperparameters for this section in the paper.
You don't need to manually register these tokens, but directly use the converted JSON for training.

**Q2&3**
Here is an evaluation code. Due to multiple modifications made to the code, I'm not sure if it's directly usable; you need to check it yourself.

```
import ast
from openai import OpenAI
import pandas as pd
import transformers
import json
from tqdm import tqdm
import concurrent.futures
import time 

def process_item(item, client, n=1):
    acc_count = 0
    messages = [
        {"role": "system", "content": item["instruction"]},
        {"role": "user", "content": item["input"]}
    ]
    
    chat_completion = client.chat.completions.create(
        messages=messages,
        model="/models/Meta-Llama-3-8B-Instruct",
        temperature=0.6,
        max_tokens=2048,
        # seed=1234,
        n=n,
        top_p = 0.9,
    )
    
    results = [choice.message.content for choice in chat_completion.choices]
    
    if item["output"] in results:
        acc_count = 1 
    return acc_count, results, item["output"]

def get_accuracy(data, client, n=1):
    total_count = 0
    results = []
    answers = []
    
    start_time = time.time()
    with concurrent.futures.ThreadPoolExecutor() as executor:
        executor._max_workers = 32
        with tqdm(total=len(data), desc="Processing") as pbar:
            futures = {executor.submit(process_item, item, client, n): item for item in data}
            for future in concurrent.futures.as_completed(futures):
                count, result, answer = future.result()
                results.append(result)
                answers.append(answer)
                total_count += count
                pbar.update(1) 
       
    end_time = time.time()       
    elapsed_time = end_time - start_time
    print(f"{elapsed_time}")
    print(f"Total elapsed time: {elapsed_time:.2f} seconds")
    print(total_count)
    
    df = pd.DataFrame({
        'results': results,
        'answers': answers
    })
    acc_at_n = total_count / len(data)
    return acc_at_n, df

div_loss = 0.25 # 0
client = OpenAI(api_key="0", base_url="http://0.0.0.0:8001/v1")
target_dataset = 'NYC'
based = 'codebook'

with open(f'/data/home/dswang/code/LLaMA-Factory/POI-RE-new/datasets/{target_dataset}/data/test_codebook_{div_loss}.json', 'r', encoding='utf-8') as file:
    data = json.load(file)
accuracy, df = get_accuracy(data, client, 1)
print("Accuracy : ", accuracy)
```
